src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,372 @@
+ */
+#ifndef MLPACK_METHODS_NE_GENOME_HPP
+#define MLPACK_METHODS_NE_GENOME_HPP
+
+#include <cstddef>
+#include <cassert>
+#include <map>
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/activation_functions/logistic_function.hpp>
+#include <mlpack/methods/ann/activation_functions/rectifier_function.hpp>
+#include <mlpack/methods/ann/activation_functions/tanh_function.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "utils.hpp"
+
+namespace mlpack {
+namespace ne {
+
+/**
+ * This class defines a genome.
+ # A genome is consist of a group of genes.
+ */
+class Genome {
+ public:
+  // Default constructor.
+  Genome() {}
+  
+  // Parametric constructor.
+  Genome(size_t id,
+  	     const std::vector<NeuronGene>& neuronGenes,
+         const std::vector<LinkGene>& linkGenes,
+         size_t numInput,
+         size_t numOutput,
+         size_t depth,
+         double fitness):
+    aId(id),
+    aNeuronGenes(neuronGenes),
+    aLinkGenes(linkGenes),
+    aNumInput(numInput),
+    aNumOutput(numOutput),
+    aDepth(depth),
+    aFitness(fitness)
+  {}
+
+  // Copy constructor.
+  Genome(const Genome& genome) {
+    aId = genome.aId;
+    aNeuronGenes = genome.aNeuronGenes;
+    aLinkGenes = genome.aLinkGenes;
+    aNumInput = genome.aNumInput;
+    aNumOutput = genome.aNumOutput;
+    aDepth = genome.aDepth;
+    aFitness = genome.aFitness;
+  }
+
+  // Destructor.
+  ~Genome() {}
+
+  // Get/set genome id.
+  size_t& Id() { return aId; }
+
+  // Get input length.
+  size_t NumInput() const { return aNumInput; }
+
+  // Set input length.
+  void NumInput(size_t numInput) { aNumInput = numInput; }
+
+  // Get output length.
+  size_t NumOutput() const { return aNumOutput; }
+
+  // Set output length.
+  void NumOutput(size_t numOutput) { aNumOutput = numOutput; }
+
+  // Get depth.
+  size_t Depth() const { return aDepth; }
+
+  // Set depth.
+  void Depth(size_t depth) { aDepth = depth; }
+
+  // Set/get fitness.
+  double& Fitness() { return aFitness; }
+
+  // Get neuron number.
+  size_t NumNeuron() const {
+    return aNeuronGenes.size();
+  }
+  
+  // Get link number.
+  size_t NumLink() const {
+    return aLinkGenes.size();
+  }
+
+  // Whether specified neuron id exist in this genome.
+  bool HasNeuronId(size_t id) const {
+    assert(id > 0);
+    assert(NumNeuron() > 0);
+
+    for (size_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+
+  // Get neuron by id.
+  NeuronGene GetNeuronById(size_t id) {
+    assert(HasNeuronId(id));
+
+    for (size_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return aNeuronGenes[i];
+      }
+    }
+  }
+
+  // Get neuron index by id.
+  size_t GetNeuronIndex(size_t id) const {
+    for(size_t i=0; i < NumNeuron(); ++i) {
+        if (aNeuronGenes[i].Id() == id) {
+            return i;
+        }
+    }
+
+    return -1;
+  }
```

Comment:
If we use size_t as return type, we can't use -1, maybe we can start with 1 and return 0?